src/nxt_conf.c

```diff
@@ -162,21 +163,24 @@ nxt_conf_json_indentation(u_char *p, uint32_t level)
 
 
 void
-nxt_conf_get_string(nxt_conf_value_t *value, nxt_str_t *str)
+nxt_conf_get_string(const nxt_conf_value_t *value, nxt_str_t *str)
 {
+    nxt_conf_value_t *v = (nxt_conf_value_t *) value;
```

Comment:
The `value` is not updated inside the function, why do we need to use the temporary `v`?

Reply:
Otherwise we get

```
... error: initialization discards ‘const’ qualifier from pointer target type [-Werror=discarded-qualifiers]
```

Which I _think_ is [this](https://port70.net/~nsz/c/c11/n1570.html#6.7.3p12)

```
pi = &cs.mem;           //   violates type constraints for =
```

Reply:
Take a look at this first.
```
static nxt_int_t
nxt_php_set_target(nxt_task_t *task, nxt_php_target_t *target,
    nxt_conf_value_t *conf)
{
    ...
    nxt_conf_value_t  *value;

    ...

    value = nxt_conf_get_object_member(conf, &root_str, NULL);

    nxt_conf_get_string(value, &str);
```
I found the function isn't changed, but if we changed the `nxt_conf_get_string()`, did you miss touching where it's called?

Reply:
I don't think the fact `value` starts off non-const before being passed into a function that takes it as a const is a problem. Indeed the compiler is perfectly happy with it and it's a perfectly reasonable thing to do, you may want to make some changes to it locally or in some other function, but _not_ some other function... E.g

```c
nxt_conf_value_t  *value;

value = ...

change_value_in_some_way(value);    // value passed in as non-const

...

dont_change_value(value);    // value passed in as const
```

True, it may be possible that in the case you showed value could be made const, but such changes were out of scope for what I was wanting to achieve  with these patches.

Reply:
> If yes, there is the same issue as above.
Sorry if I misunderstand you. It's not an issue. I mean the new way introduced a bit of confusion.
Let's return the function.
```
void
nxt_conf_get_string(nxt_conf_value_t *value, nxt_str_t *str)
{
    if (value->type == NXT_CONF_VALUE_SHORT_STRING) {
        str->length = value->u.str.length;
        str->start = value->u.str.start;

    } else {
        str->length = value->u.string.length;
        str->start = value->u.string.start;
    }
}
```
It's to get the string from the conf value.
The key is if `str` is changeable. I feel it's read-only, if yes, it should have its own allocation, for example:
```
void
nxt_conf_get_string(nxt_conf_value_t *value, nxt_str_t *str)
{
    ...
    if (value->type == NXT_CONF_VALUE_SHORT_STRING) {
        length = value->u.str.length;
        start = value->u.str.start;

    } else {
        length = value->u.string.length;
        start = value->u.string.start;
    }

    str->start = malloc(length); /* mp should be used */
    ...
}
```

So if we make `value` const, it seems `str` is also const. For example:
```
void
nxt_conf_get_string(const nxt_conf_value_t *value, const nxt_str_t *str)
```

Reply:
Hmm, wait, hang on.

I'm maybe not understanding what you're getting at at, but it won't make any difference if value is const or not before calling `nxt_conf_get_string()`, you'd still get the compiler warning...

Reply:
1. The following is a bit confusing. Reviewers need to understand it's because of the compiler warning.
```
nxt_conf_value_t *v = (nxt_conf_value_t *) value;
```
2. It seems it's unnecessary to add the above to avoid warning, and it's better not to add something like it.
3. If `value` is const, `str` is also const, since they are read-only and str gets part from `value`.

Reply:
If `str` is `const`, then we can't set any of its members. E.g quick hack to demonstrate

```c
void                                                                            
nxt_conf_get_string(const nxt_conf_value_t *value, const nxt_str_t *str)        
{                                                                                                                                               
    if (value->type == NXT_CONF_VALUE_SHORT_STRING) {                           
        str->length = value->u.str.length;                                      
        str->start = value->u.str.start;                                        
                                                                                
    } else {                                                                    
        str->length = value->u.string.length;                                   
        str->start = value->u.string.start;                                     
    }                                                                           
} 
```

```
  CC     build/src/nxt_conf.o
src/nxt_conf.c: In function ‘nxt_conf_get_string’:
src/nxt_conf.c:169:21: error: assignment of member ‘length’ in read-only object
  169 |         str->length = value->u.str.length;
      |                     ^
src/nxt_conf.c:170:20: error: assignment of member ‘start’ in read-only object
  170 |         str->start = value->u.str.start;
      |                    ^
src/nxt_conf.c:173:21: error: assignment of member ‘length’ in read-only object
  173 |         str->length = value->u.string.length;
      |                     ^
src/nxt_conf.c:174:20: error: assignment of member ‘start’ in read-only object
  174 |         str->start = value->u.string.start;
      |                    ^
```

Of course you'd hit the exact same thing with the `str->start = malloc(length);`

This is the point of making things const after all...

The only way I can immediately think of to make value const and not have the cast is by using memcpy(3)...

Reply:
> If str is const, then we can't set any of its members.

I didn't test it, it's to show the idea. Anyway, if we touch the function, it's important to check if `str` is read-only.

Reply:
Will check it tomorrow, not sure if my idea is correct.
The same for the two changes you dropped, they might make sense.

Reply:
Btw, you could try with this.
```
void
nxt_conf_get_string(const nxt_conf_value_t *value, nxt_str_t *str)
{
    if (value->type == NXT_CONF_VALUE_SHORT_STRING) {
        str->length = value->u.str.length;
        str->start = (u_char *) value->u.str.start;

    } else {
        str->length = value->u.string.length;
        str->start = value->u.string.start;
    }
}
```

Reply:
> I didn't test it, it's to show the idea. Anyway, if we touch the function, it's important to check if str is read-only.

It _can't_ be read-only as we _need_ to modify it...

Indeed, your suggestion works.

For some bizarre reason I thought it was failing on the `start->length = `, which didn't make sense to me, no wonder!

So you're happy with `(u_char *)` casts?

As the dropped changes, I'm happy to do them in a separate commit anyway and am not even 100% conviced by the `nxt_conf_get_path()` one...

Reply:
> ```
> nxt_conf_value_t *
> nxt_conf_get_path(const nxt_conf_value_t *value, nxt_str_t *path)
> ```
>
> I feel the declaration is correct. And can this work without adding v?
> ```
> @@ -510,7 +510,7 @@ nxt_conf_get_path(nxt_conf_value_t *value, nxt_str_t *path)
>
>    } while (parse.last == 0);
>
> -    return value;
> +    return (nxt_conf_value_t *) value;
>  }
> ```

Yes, that is actually what I had originally...

```diff
@@ -459,7 +459,7 @@ nxt_conf_type(nxt_conf_value_t *value)
 
 
 nxt_conf_value_t *
-nxt_conf_get_path(nxt_conf_value_t *value, nxt_str_t *path)
+nxt_conf_get_path(const nxt_conf_value_t *value, const nxt_str_t *path)
 {
     nxt_str_t              token;
     nxt_int_t              ret, index;
@@ -510,7 +510,7 @@ nxt_conf_get_path(nxt_conf_value_t *value, nxt_str_t *path)
 
     } while (parse.last == 0);
 
-    return value;
+    return (nxt_conf_value_t *) value;
 }
 
 
```

The bit I'm not 100% sure about is, that yes, while it clearly works, taking a const then returning _it_ as a non-const perhaps has a slight _smell_ to it... need to mull it over...